File: engine/swiftui/Sources/MilanoSDK/Engine/MilanoRegistry.swift
import Foundation

/// The binding from vocabulary component types to consumer renderers.
public struct MilanoRegistry {
    private(set) var renderers: [String: MilanoRenderer] = [:]
    private(set) var placeholder: MilanoPlaceholderRenderer?

    public init() {}

    /// Registers a renderer for one component type name.
    public mutating func register(_ renderer: MilanoRenderer, for componentType: String) {
        renderers[componentType] = renderer
    }

    /// Registers the placeholder renderer, required only when the
    /// unknown-type policy is `.placeholder`.
    public mutating func registerPlaceholder(_ renderer: MilanoPlaceholderRenderer) {
        placeholder = renderer
    }
}

File: engine/compose/src/commonMain/kotlin/dev/getmilano/engine/MilanoRegistry.kt
package dev.getmilano

/** The binding from vocabulary component types to consumer renderers. */
class MilanoRegistry {
    internal val renderers = LinkedHashMap<String, MilanoRenderer>()
    internal var placeholder: MilanoPlaceholderRenderer? = null
        private set

    /** Registers a renderer for one component type name. */
    fun register(
        componentType: String,
        renderer: MilanoRenderer,
    ) {
        renderers[componentType] = renderer
    }

    /**
     * Registers the placeholder renderer, required only when the
     * unknown-type policy is [MilanoUnknownTypePolicy.PLACEHOLDER].
     */
    fun registerPlaceholder(renderer: MilanoPlaceholderRenderer) {
        placeholder = renderer
    }
}
